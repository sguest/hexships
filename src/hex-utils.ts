// https://www.redblobgames.com/grids/hexagons/
const sqrt3 = Math.sqrt(3);

export function getCenter(x: number, y: number, size: number) {
    return {
        x: 1.5 * x * size,  // 0.75 * 2 * x * size
        y: (y + 0.5 * x) * size * sqrt3,
    }
}

export function getCorners(x: number, y: number, size: number) {
    const center = getCenter(x, y, size);
    return [...Array(6).keys()].map(i => {
        const angle = Math.PI / 3 * i;  // Math.PI / 180 * 60 * i === i * 60deg converted to radians
        return {
            x: center.x + size * Math.cos(angle),
            y: center.y + size * Math.sin(angle),
        }
    })
}

export function isInGrid(x: number, y: number, gridSize: number) {
    const z = -x - y;
    return x > -gridSize
        && x < gridSize
        && y > -gridSize
        && y < gridSize
        && z > -gridSize
        && z < gridSize;
}

export function getGridCells(gridSize: number) {
    const cells: Array<{x: number, y: number}> = [];

    for(let x = -gridSize + 1; x < gridSize; x++) {
        for(let y = -gridSize + 1; y < gridSize; y++) {
            if(isInGrid(x, y, gridSize)) {
                cells.push({x, y});
            }
        }
    }

    return cells;
}