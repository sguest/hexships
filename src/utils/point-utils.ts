export interface Point {
    x: number
    y: number
}

export function add(...points: Point[]) {
    const point = { x: 0, y: 0 };
    for(let p of points) {
        point.x += p.x;
        point.y += p.y;
    }
    return point;
}

export function multiplyScalar(point: Point, scalar: number) {
    return {
        x: point.x * scalar,
        y: point.y * scalar,
    }
}

export function equal(a: Point, b: Point) {
    return a.x === b.x && a.y === b.y;
}

export function moveTo(context: CanvasRenderingContext2D, point: Point) {
    context.moveTo(point.x, point.y);
}

export function lineTo(context: CanvasRenderingContext2D, point: Point) {
    context.lineTo(point.x, point.y);
}