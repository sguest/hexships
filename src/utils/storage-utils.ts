export function encodeLocalStorage(key: string, value: any) {
    const encoded = window.btoa(encodeURIComponent(JSON.stringify(value)));
    localStorage.setItem(key, encoded);
}

export function decodeLocalStorage(key: string) {
    try {
        const encoded = localStorage.getItem(key);
        if(!encoded?.length) {
            return encoded;
        }
        return JSON.parse(decodeURIComponent(window.atob(encoded)));
    }
    catch {
        // Invalid storage value. Not interested in logging, just return null
        return null;
    }
}