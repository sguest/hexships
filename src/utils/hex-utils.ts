import { Point } from './point-utils';

// https://www.redblobgames.com/grids/hexagons/
const sqrt3 = Math.sqrt(3);

export function getCenter(cell: Point, size: number) {
    return {
        x: 1.5 * cell.x * size, // 0.75 * 2 * x * size
        y: (cell.y + 0.5 * cell.x) * size * sqrt3,
    }
}

export function getCorners(cell: Point, size: number) {
    const center = getCenter(cell, size);
    return [...Array(6).keys()].map(i => {
        const angle = Math.PI / 3 * i; // Math.PI / 180 * 60 * i === i * 60deg converted to radians
        return {
            x: center.x + size * Math.cos(angle),
            y: center.y + size * Math.sin(angle),
        }
    })
}

export function roundFractional(coords: Point) {
    const zCoord = -coords.x - coords.y;
    let x = Math.round(coords.x);
    let y = Math.round(coords.y);
    const z = Math.round(zCoord);

    const xDiff = Math.abs(x - coords.x);
    const yDiff = Math.abs(y - coords.y);
    const zDiff = Math.abs(z - zCoord);

    if(xDiff > yDiff && xDiff > zDiff) {
        x = -y - z;
    }
    else if(yDiff > zDiff) {
        y = -x - z;
    }

    return { x, y };
}

export function getCellFromCoords(hit: Point, size: number) {
    return roundFractional({
        x: hit.x * 2 / 3 / size,
        y: (hit.x * -1 / 3 + hit.y * sqrt3 / 3) / size,
    });
}

export function isInGrid(cell: Point, gridSize: number) {
    const z = -cell.x - cell.y;
    return cell.x > -gridSize &&
        cell.x < gridSize &&
        cell.y > -gridSize &&
        cell.y < gridSize &&
        z > -gridSize &&
        z < gridSize;
}

export function getGridCells(gridSize: number) {
    const cells: Array<{x: number, y: number}> = [];

    for(let x = -gridSize + 1; x < gridSize; x++) {
        for(let y = -gridSize + 1; y < gridSize; y++) {
            if(isInGrid({ x, y }, gridSize)) {
                cells.push({ x, y });
            }
        }
    }

    return cells;
}