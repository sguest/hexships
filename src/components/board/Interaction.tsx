import React, { useEffect, useRef, useState } from 'react';
import { Point } from '../../utils/point-utils';
import * as hexUtils from '../../utils/hex-utils';
import * as pointUtils from '../../utils/point-utils';
import useScaledCanvas from './useScaledCanvas';
import { createUseStyles } from 'react-jss';

export interface InteractionProps {
    highlightTiles?: Array<{x: number, y: number, style: string | CanvasGradient | CanvasPattern}>
    mouseHighlightStyle?: (tile: Point) => string | CanvasGradient | CanvasPattern | undefined
    gridSize: number
    uiScale: number
    gridDimensions: Point
    onSelectTile?: (tile: Point) => void
}

const useStyles = createUseStyles({
    canvas: {
        position: 'absolute',
        zIndex: 2,
    },
})

export default function Interaction(props: InteractionProps) {
    const canvasRef = useRef<HTMLCanvasElement>(null);
    const [hoverTile, setHoverTile] = useState<Point | null>(null);
    const [hoverStyle, setHoverStyle] = useState<string | CanvasGradient | CanvasPattern | undefined>(undefined);
    const classes = useStyles();

    useEffect(() => {
        useScaledCanvas(canvasRef, props.uiScale, context => {
            const fillTile = (tile: Point, style: string | CanvasGradient | CanvasPattern) => {
                const corners = hexUtils.getCorners(tile);
                context.fillStyle = style;
                context.beginPath();
                pointUtils.moveTo(context, pointUtils.add(corners[5], pointUtils.multiplyScalar(props.gridDimensions, 0.5)));
                for(const corner of corners) {
                    pointUtils.lineTo(context, pointUtils.add(corner, pointUtils.multiplyScalar(props.gridDimensions, 0.5)));
                }
                context.fill();
            }

            if(hoverTile && hoverStyle) {
                fillTile(hoverTile, hoverStyle);
            }

            if(props.highlightTiles) {
                for(const tile of props.highlightTiles) {
                    fillTile(tile, tile.style);
                }
            }
        });
    }, [canvasRef, props.highlightTiles, hoverTile, hoverStyle, props.uiScale, props.gridDimensions])

    const getMouseTile = (e: React.MouseEvent<HTMLCanvasElement>) => {
        const rect = (e.target as HTMLCanvasElement).getBoundingClientRect();
        const relativeClick = pointUtils.add(
            pointUtils.multiplyScalar({ x: e.clientX, y: e.clientY }, 1 / props.uiScale),
            pointUtils.multiplyScalar(props.gridDimensions, -0.5),
            pointUtils.multiplyScalar(rect, -1 / props.uiScale),
        );
        return hexUtils.getCellFromCoords(relativeClick);
    }

    const onClick = (e: React.MouseEvent<HTMLCanvasElement>) => {
        const clickedTile = getMouseTile(e);
        if(hexUtils.isInGrid(clickedTile, props.gridSize)) {
            props.onSelectTile && props.onSelectTile(clickedTile);
        }
    }

    const onMove = (e: React.MouseEvent<HTMLCanvasElement>) => {
        const tile = getMouseTile(e);
        setHoverTile(hexUtils.isInGrid(tile, props.gridSize) ? tile : null);
        setHoverStyle(props.mouseHighlightStyle ? props.mouseHighlightStyle(tile) : undefined);
    }

    const onLeave = () => {
        setHoverTile(null);
    }

    return <canvas ref={canvasRef}
        width={props.gridDimensions.x * props.uiScale}
        height={props.gridDimensions.y * props.uiScale}
        className={classes.canvas}
        onClick={onClick}
        onMouseMove={onMove}
        onMouseLeave={onLeave} />
}