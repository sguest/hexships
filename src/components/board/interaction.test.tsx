import 'jest-canvas-mock';
import { render, screen, fireEvent } from '@testing-library/react';
import Interaction from './Interaction';
import * as hexUtils from '../../utils/hex-utils';

test('should render highlight tile', () => {
    render(<Interaction highlightTiles={[{ x: 1, y: 1, style: 'orange' }]} gridSize={7} uiScale={1} gridDimensions={{ x: 1, y: 1 }} />)
    const canvas = screen.getByTestId('interaction-canvas') as HTMLCanvasElement;
    const context = canvas.getContext('2d');
    const events = context!.__getEvents();
    expect(events).toMatchSnapshot();
});

test('should render overlay', () => {
    render(<Interaction overlayStyle="green" gridSize={7} uiScale={1} gridDimensions={{ x: 1, y: 1 }} />)
    const canvas = screen.getByTestId('interaction-canvas') as HTMLCanvasElement;
    const context = canvas.getContext('2d');
    const events = context!.__getEvents();
    expect(events).toMatchSnapshot();
});

test('should highlight mouseover tile', () => {
    render(<Interaction mouseHighlightStyle={() => 'yellow'} gridSize={7} uiScale={1} gridDimensions={{ x: 1, y: 1 }} />)
    const canvas = screen.getByTestId('interaction-canvas') as HTMLCanvasElement;
    fireEvent.mouseMove(canvas, { clientX: 50, clientY: 20 });

    const context = canvas.getContext('2d');
    const events = context!.__getEvents();
    expect(events).toMatchSnapshot();
});

test('should execute callback on tile click', () => {
    const callback = jest.fn();
    render(<Interaction onSelectTile={callback} gridSize={7} uiScale={1} gridDimensions={{ x: 1, y: 1 }} />)
    const canvas = screen.getByTestId('interaction-canvas') as HTMLCanvasElement;
    const mousePos = { x: 50, y: 20 };
    fireEvent.click(canvas, { clientX: mousePos.x, clientY: mousePos.y });
    expect(callback).toHaveBeenCalledWith(hexUtils.getCellFromCoords(mousePos));
});