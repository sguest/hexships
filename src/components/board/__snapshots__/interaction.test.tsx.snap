// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should highlight mouseover tile 1`] = `
Array [
  Object {
    "props": Object {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "value": "#ffff00",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 70.49999999999999,
      "y": 17.82050807568876,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 80.5,
      "y": 35.14101615137754,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 70.5,
      "y": 52.46152422706631,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 50.5,
      "y": 52.46152422706632,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 40.5,
      "y": 35.14101615137754,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 50.49999999999999,
      "y": 17.82050807568877,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 70.49999999999999,
      "y": 17.82050807568876,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 70.49999999999999,
            "y": 17.82050807568876,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 80.5,
            "y": 35.14101615137754,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 70.5,
            "y": 52.46152422706631,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 50.5,
            "y": 52.46152422706632,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 40.5,
            "y": 35.14101615137754,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 50.49999999999999,
            "y": 17.82050807568877,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 70.49999999999999,
            "y": 17.82050807568876,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`should render highlight tile 1`] = `
Array [
  Object {
    "props": Object {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "value": "#ffa500",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 40.499999999999986,
      "y": 35.14101615137753,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 50.5,
      "y": 52.46152422706631,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 40.5,
      "y": 69.78203230275508,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 20.500000000000004,
      "y": 69.78203230275508,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 10.5,
      "y": 52.46152422706631,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 20.499999999999993,
      "y": 35.14101615137754,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "x": 40.499999999999986,
      "y": 35.14101615137753,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 40.499999999999986,
            "y": 35.14101615137753,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 50.5,
            "y": 52.46152422706631,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 40.5,
            "y": 69.78203230275508,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 20.500000000000004,
            "y": 69.78203230275508,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 10.5,
            "y": 52.46152422706631,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 20.499999999999993,
            "y": 35.14101615137754,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
        Object {
          "props": Object {
            "x": 40.499999999999986,
            "y": 35.14101615137753,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
]
`;

exports[`should render overlay 1`] = `
Array [
  Object {
    "props": Object {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
  Object {
    "props": Object {
      "height": 1,
      "width": 1,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "x": 1,
      "y": 1,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "scale",
  },
  Object {
    "props": Object {
      "value": "#008000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 999999,
      "width": 999999,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;
